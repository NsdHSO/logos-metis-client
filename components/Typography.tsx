import { Text, TextProps } from "react-native";
import { web } from "@app/platform/detection";
import { atoms, HeadingLevels, headingStyles } from "@app/util/atoms";
import React from "react";

export interface Typography {}

/**
 * Our main text component. Use this most of the time.
 */
export function TextSyn({ children, style, selectable, ...rest }: TextProps) {
  const shared = {
    uiTextView: true,
    selectable,
    style: style,
    dataSet: Object.assign({}),
    ...rest,
  };

  return <Text {...shared}>{children}</Text>;
}

// Function to map heading level to the style
export function createHeadingElement({ level }: { level: HeadingLevels }) {
  return function HeadingElement({ style, ...rest }: TextProps) {
    const headingStyle = headingStyles[level]; // Get the corresponding style
    const attr = web({ role: "heading", "aria-level": level }) || {};

    return (
      <TextSyn
        {...attr}
        {...rest}
        style={[style, headingStyle]} // Apply the mapped style
      />
    );
  };
}

/*
 * Use semantic components when it's beneficial to the user or to a web scraper
 */
export const H1 = createHeadingElement({ level: 1 });
export const H2 = createHeadingElement({ level: 2 });
export const H3 = createHeadingElement({ level: 3 });
export const H4 = createHeadingElement({ level: 4 });
export const H5 = createHeadingElement({ level: 5 });
export const H6 = createHeadingElement({ level: 6 });
export const H7 = createHeadingElement({ level: 7 });
export const H8 = createHeadingElement({ level: 8 });

export function P({ style, ...rest }: TextProps) {
  const attr =
    web({
      role: "paragraph",
    }) || {};
  return <TextSyn {...attr} {...rest} style={[atoms.text_md, style]} />;
}
