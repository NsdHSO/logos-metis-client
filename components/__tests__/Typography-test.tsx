import * as React from "react";
import renderer from "react-test-renderer";
import { H1, H2, H3, H4, H5, H6, H7, H8, P } from '@components/Typography';

it(`renders correctly`, () => {
  const tree = renderer.create(<H1>Snapshot test!</H1>).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should have prop', () => {
  const tree = renderer.create(<H1>Iancu</H1>).toJSON()as any;
  expect(tree).toMatchSnapshot();

  expect(...tree.children).toEqual('Iancu')
  expect(tree.props.style[1][1].fontSize).toEqual(40)
});


it('should have prop h2', () => {
  const tree = renderer.create(<H2>Iancu</H2>).toJSON()as any;
  expect(tree).toMatchSnapshot();

  expect(...tree.children).toEqual('Iancu')
  expect(tree.props.style[1][1].fontSize).toEqual(32)
});


it('should have prop h3', () => {
  const tree = renderer.create(<H3>Iancu</H3>).toJSON()as any;
  expect(tree).toMatchSnapshot();

  expect(...tree.children).toEqual('Iancu')
  expect(tree.props.style[1][1].fontSize).toEqual(26)
});


it('should have prop h4', () => {
  const tree = renderer.create(<H4>Iancu</H4>).toJSON()as any;
  expect(tree).toMatchSnapshot();

  expect(...tree.children).toEqual('Iancu')
  expect(tree.props.style[1][1].fontSize).toEqual(22)
});


it('should have prop h5', () => {
  const tree = renderer.create(<H5>Iancu</H5>).toJSON()as any;
  expect(tree).toMatchSnapshot();

  expect(...tree.children).toEqual('Iancu')
  expect(tree.props.style[1][1].fontSize).toEqual(20)
});


it('should have prop h6', () => {
  const tree = renderer.create(<H6>Iancu</H6>).toJSON()as any;
  expect(tree).toMatchSnapshot();

  expect(...tree.children).toEqual('Iancu')
  expect(tree.props.style[1][1].fontSize).toEqual(18)
});


it('should have prop h7', () => {
  const tree = renderer.create(<H7>Iancu</H7>).toJSON()as any;
  expect(tree).toMatchSnapshot();

  expect(...tree.children).toEqual('Iancu')
  expect(tree.props.style[1][1].fontSize).toEqual(16)
});


it('should have prop h8', () => {
  const tree = renderer.create(<H8>Iancu</H8>).toJSON()as any;
  expect(tree).toMatchSnapshot();

  expect(...tree.children).toEqual('Iancu')
  expect(tree.props.style[1][1].fontSize).toEqual(14)
});

it('should have prop p', () => {
  const tree = renderer.create(<P>Iancu</P>).toJSON()as any;
  expect(tree).toMatchSnapshot();

  expect(...tree.children).toEqual('Iancu')
  expect(tree.props.style[1][0].fontSize).toEqual(16)
});