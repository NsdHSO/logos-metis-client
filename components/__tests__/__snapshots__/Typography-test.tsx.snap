// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Text
  dataSet={{}}
  style={
    [
      {
        "color": "#11181C",
      },
      [
        undefined,
        {
          "fontSize": 40,
          "fontWeight": "600",
          "letterSpacing": 1.625,
          "lineHeight": 40,
        },
      ],
    ]
  }
  uiTextView={true}
>
  Snapshot test!
</Text>
`;

exports[`should have prop 1`] = `
<Text
  dataSet={{}}
  style={
    [
      {
        "color": "#11181C",
      },
      [
        undefined,
        {
          "fontSize": 40,
          "fontWeight": "600",
          "letterSpacing": 1.625,
          "lineHeight": 40,
        },
      ],
    ]
  }
  uiTextView={true}
>
  Iancu
</Text>
`;

exports[`should have prop h1 1`] = `
<Text
  dataSet={{}}
  style={
    [
      {
        "color": "#11181C",
      },
      [
        undefined,
        {
          "fontSize": 32,
          "fontWeight": "600",
          "letterSpacing": 0,
        },
      ],
    ]
  }
  uiTextView={true}
>
  Iancu
</Text>
`;

exports[`should have prop h2 1`] = `
<Text
  dataSet={{}}
  style={
    [
      {
        "color": "#11181C",
      },
      [
        undefined,
        {
          "fontSize": 32,
          "fontWeight": "600",
          "letterSpacing": 0,
        },
      ],
    ]
  }
  uiTextView={true}
>
  Iancu
</Text>
`;

exports[`should have prop h3 1`] = `
<Text
  dataSet={{}}
  style={
    [
      {
        "color": "#11181C",
      },
      [
        undefined,
        {
          "fontSize": 26,
          "fontWeight": "600",
          "letterSpacing": 0,
        },
      ],
    ]
  }
  uiTextView={true}
>
  Iancu
</Text>
`;

exports[`should have prop h4 1`] = `
<Text
  dataSet={{}}
  style={
    [
      {
        "color": "#11181C",
      },
      [
        undefined,
        {
          "fontSize": 22,
          "fontWeight": "600",
          "letterSpacing": 0,
        },
      ],
    ]
  }
  uiTextView={true}
>
  Iancu
</Text>
`;

exports[`should have prop h5 1`] = `
<Text
  dataSet={{}}
  style={
    [
      {
        "color": "#11181C",
      },
      [
        undefined,
        {
          "fontSize": 20,
          "fontWeight": "600",
          "letterSpacing": 0,
        },
      ],
    ]
  }
  uiTextView={true}
>
  Iancu
</Text>
`;

exports[`should have prop h6 1`] = `
<Text
  dataSet={{}}
  style={
    [
      {
        "color": "#11181C",
      },
      [
        undefined,
        {
          "fontSize": 18,
          "fontWeight": "600",
          "letterSpacing": 0,
        },
      ],
    ]
  }
  uiTextView={true}
>
  Iancu
</Text>
`;

exports[`should have prop h7 1`] = `
<Text
  dataSet={{}}
  style={
    [
      {
        "color": "#11181C",
      },
      [
        undefined,
        {
          "fontSize": 16,
          "fontWeight": "600",
          "letterSpacing": 0,
        },
      ],
    ]
  }
  uiTextView={true}
>
  Iancu
</Text>
`;
